### DISCUSSION TITLE
Protocol extension to BMS JK-B2A24S15P

### DISCUSSION THREAD

**Discussion by Holric** [jblance/mpp-solar#72] [General]
Hi,
I did finish the general setup of my solar battery logging using mpp-solar and Grafana.
Two BMS JK-B2A24S15P (combined BMS and active Balancer) and a Morningstar Tristar MPPT are logged via Telegraf using Json.
The actual mpp-solar is working fine, using configuration options as filter and exclfilter is very comfortable now.

I made a first progress in finding and adding some additional JK-B2A24S15P specific fields in the protocol. I used the former JK02.py containing all the unknowns.
Maybe it would be useful to add a separate protocol for the combined BMS / balancers if it interferes with the active balancers without BMS.
Changed JK02 response part:

        "response": [
            ["hex", 4, "Header", ""],
            ["hex", 1, "Record_Type", ""],
            ["int", 1, "Record_Counter", ""],
            ["loop", 24, "Voltage_Cell", "V", "2ByteHex"],
            ["discard", 4, "discard1", ""],
            ["2ByteHex", 1, "Average_Cell_Voltage", "V"],
            ["2ByteHex", 1, "Delta_Cell_Voltage", "V"],
            ["2ByteHexU", 1, "Unknown1", ""],
            ["loop", 25, "Resistance_Cell", "Ohm", "2ByteHex"],
            ["discard", 4, "discard2", ""],
            ["2ByteHex", 1, "Battery_Voltage", "V"],
            ["discard", 10, "discard3", ""],
            ["2ByteHexC", 1, "Battery_T1", "°C"],
            ["2ByteHexC", 1, "Battery_T2", "°C"],
            ["2ByteHexC", 1, "MOS_Temp", "°C"],
            ["discard", 4, "discard4", ""],
            ["discard", 1, "discard4_1", ""], #added
            ["int", 1, "Percent_Remain", ""], #Unknown5
            ["2ByteHexU", 1, "Capacity_Remain", ""], #Unknown6
            ["2ByteHexU", 1, "Unknown7", ""],
            ["discard", 1, "Capacity", ""], #Unknown8 Filled in Battery capacity, output as string, maybe Unknown7 is related if > 99Ah
            ["discard", 1, "discard4_2", ""], #added
            ["2ByteHexU", 1, "Unknown9", ""],
            ["discard", 4, "discard5", ""],
            ["2ByteHexU", 1, "Unknown10", ""],
            ["2ByteHexU", 1, "Unknown11", ""],
            ["2ByteHexU", 1, "Unknown12", ""],
            ["2ByteHexU", 1, "Unknown13", ""],
            ["uptime", 3, "Time", ""],
            ["2ByteHexU", 1, "Unknown15", ""],
            ["2ByteHexU", 1, "Unknown16", ""],
            ["2ByteHexU", 1, "Unknown17", ""],
            ["discard", 12, "discard6", ""],
            ["2ByteHexU", 1, "Unknown18", ""],
            ["2ByteHexU", 1, "Unknown19", ""],
            ["2ByteHexU", 1, "Unknown20", ""],
            ["2ByteHexU", 1, "Current_Charge", ""], #Unknown21
            ["2ByteHexU", 1, "Current_Discharge", ""], #Unknown22
            ["2ByteHexU", 1, "Unknown23", ""],
            ["2ByteHexU", 1, "Unknown24", ""],
            ["2ByteHexU", 1, "Unknown25", ""],
            ["2ByteHexU", 1, "Unknown26", ""],
            ["2ByteHexU", 1, "Unknown27", ""],
            ["2ByteHexU", 1, "Unknown28", ""],
            ["2ByteHexU", 1, "Unknown29", ""],
            ["rem"],
        ],
I have a 80Ah and 90Ah battery, found it in left/first byte of Unknown8 as "80" and "90" and could not see, whether Unknown7 and first byte Unknown8 are related if Battery capacity is more than 99Ah. I will keep on finding out more.
Maybe there is some interest to get all available information out of the combined BMS / balancers.

For logging the Tristar MPPT solar charger, I did modify the source from https://github.com/ajay10000/tristar_mppt to get a JSON output for telegraf. If there is interest in using it or integrating it into mpp-solar, I can upload it.

**Comment by Holric** [comment]
I was a little bit too hasty, while getting no proper results on all capacity values I found out the reason.
Some values are using 4 hex bytes coded as 32bit int / 1000.
I added a conversion for this case and had some success:

        "response": [
            ["hex", 4, "Header", ""],
            ["hex", 1, "Record_Type", ""],
            ["int", 1, "Record_Counter", ""],
            ["loop", 24, "Voltage_Cell", "V", "2ByteHex"],
            ["discard", 4, "discard1", ""],
            ["2ByteHex", 1, "Average_Cell_Voltage", "V"],
            ["2ByteHex", 1, "Delta_Cell_Voltage", "V"],
            ["2ByteHex", 1, "Current_Balancer", ""],
            ["loop", 25, "Resistance_Cell", "Ohm", "2ByteHex"],
            ["discard", 4, "discard2", ""],
            ["2ByteHex", 1, "Battery_Voltage", "V"],
            ["discard", 10, "discard3", ""],
            ["2ByteHexC", 1, "Battery_T1", "°C"],
            ["2ByteHexC", 1, "Battery_T2", "°C"],
            ["2ByteHexC", 1, "MOS_Temp", "°C"],
            ["discard", 4, "discard4", ""], #discard4
            ["discard", 1, "discard4_1", ""], #added
            ["int", 1, "Percent_Remain", ""], #Unknown5
            ["4ByteHex1000", 4, "Capacity_Remain", ""], #Unknown6+7
            ["4ByteHex1000", 4, "Nominal_Capacity", ""], #Unknown8+9
            ["discard", 4, "discard5", ""],
            #["discard", 2, "Unknown10", ""],
            #["discard", 2, "Unknown11", ""],
            ["4ByteHex1000", 4, "Capacity_Cycle", ""], #Unknown10+11
            ["discard", 2, "Unknown12", ""],
            ["discard", 2, "Unknown13", ""],
            ["uptime", 3, "Time", ""],
            ["discard", 2, "Unknown15", ""],
            ["discard", 2, "Unknown16", ""],
            ["discard", 2, "Unknown17", ""],
            ["discard", 12, "discard6", ""],
            ["discard", 2, "Unknown18", ""],
            ["discard", 2, "Unknown19", ""],
            ["discard", 2, "Unknown20", ""],
            ["2ByteHex", 1, "Current_Charge", ""], #Unknown21
            ["2ByteHex", 1, "Current_Discharge", ""], #Unknown22
            ["discard", 2, "Unknown23", ""],
            ["discard", 2, "Unknown24", ""],
            ["discard", 2, "Unknown25", ""],
            ["discard", 2, "Unknown26", ""],
            ["discard", 2, "Unknown27", ""],
            ["discard", 2, "Unknown28", ""],
            ["discard", 2, "Unknown29", ""],
            ["rem"],

in jkabstractprotocol.py:

                elif defn[0] == "4ByteHex1000":
                    log.debug("decode: 4ByteHex1000 defn")
                    v = responses[:4]
                    responses = responses[4:]
                    value = decode4ByteHex1000(v)
                    msgs[defn[2]] = [value, defn[3]]

in protocol_helpers.py:

 def decode4ByteHex1000(hexToDecode):

    hexString = hexToDecode
    log.debug(f"hexString: {hexString}")

    answer = 0.0

    # Make sure supplied String is long enough
    if len(hexString) != 4:
        log.warning(f"Hex encoded value must be 4 bytes long. Was {len(hexString)} length")
        return 0

    # 1st position
    pos1 = hexString[0] >> 4
    answer += pos1 * (2 ** 4 / 1000)
    log.debug(f"answer after pos1 {answer}")
    # 2nd position
    pos2 = hexString[0] & 0x0F
    answer += pos2 * (2 ** 0 / 1000)
    log.debug(f"answer after pos2 {answer}")
    # 3rd position
    pos3 = hexString[1] >> 4
    answer += pos3 * (2 ** 12 / 1000)
    log.debug(f"answer after pos3 {answer}")
    # 4th position
    pos4 = hexString[1] & 0x0F
    answer += pos4 * (2 ** 8 / 1000)
    # 5st position
    pos5 = hexString[2] >> 4
    answer += pos5 * (2 ** 20 / 1000)
    log.debug(f"answer after pos5 {answer}")
    # 6st position
    pos6 = hexString[2] & 0x0F
    answer += pos6 * (2 ** 16 / 1000)
    log.debug(f"answer after pos6 {answer}")
    # 7th position
    pos7 = hexString[3] >> 4
    answer += pos7 * (2 ** 28 / 1000)
    log.debug(f"answer after pos7 {answer}")
    # 8th position
    pos8 = hexString[3] & 0x0F
    answer += pos8 * (2 ** 24 / 1000)

    log.debug(f"answer after pos8 {answer}")
    log.info(f"Hex {hexString} 8 byte decoded to {answer}")

    return answer

I attached the modified files. 
The battery remaining and nominal capacity values, battery capacity in percent  and the cycle capacity are working now.

[mpp-solar-changed_protocol.zip](https://github.com/jblance/mpp-solar/files/6004344/mpp-solar-changed_protocol.zip)

**Reply by jblance** [reply]
OK, thanks
These look to be combatible to non-active balancers as well - so i have added these changes to JK02
